langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/internal/MatchingService.java

```diff
@@ -0,0 +1,111 @@
+     * @return the list of matching documents
+     */
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        final IndexDatapoint datapoint = IndexDatapoint
+                .newBuilder()
+                .addAllFeatureVector(referenceEmbedding.vectorAsList())
+                .build();
+        final FindNeighborsRequest.Query query = FindNeighborsRequest.Query
+                .newBuilder()
+                .setDatapoint(datapoint)
+                .setNeighborCount(maxResults)
+                .build();
+        final FindNeighborsRequest request = FindNeighborsRequest
+                .newBuilder()
+                .setIndexEndpoint(IndexEndpointName.of(project, location, indexEndpointId).toString())
+                .setDeployedIndexId(deployedIndexId)
+                .addQueries(query)
+                .setReturnFullDatapoint(returnFullDatapoint)
+                .build();
+
+        try (MatchServiceClient client = MatchServiceClient.create(getSettings())) {
+            final FindNeighborsResponse response = client.findNeighbors(request);
+            return response.getNearestNeighborsList()
+                    .stream()
+                    .flatMap(neighbor -> neighbor.getNeighborsList().stream())
+                    .filter(neighbor -> neighbor.getDistance() >= minScore)
+                    .map(neighbor -> {
+                        final String id = neighbor.getDatapoint().getDatapointId();
+                        final String path = "documents/" + id;
+                        final String content = gcpBlobService.download(path);
+                        final VertexAIDocument document = VertexAIDocument.fromJson(content);
+                        final IndexDatapoint resultDatapoint = neighbor.getDatapoint();
+                        return new EmbeddingMatch<>(neighbor.getDistance(),
+                                id,
+                                Embedding.from(resultDatapoint.getFeatureVectorList()),
+                                (document != null) ? document.getSegment() : null);
+                    })
+                    .collect(Collectors.toList());
+        } catch (Exception e) {
+            log.error("Error finding neighbors", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Initialize the Vertex AI MatchServiceSettings
+     *
+     * @return the initialized MatchServiceSettings
+     */
+    private MatchServiceSettings initSettings() {
+        try {
+            final MatchServiceSettings.Builder settings = MatchServiceSettings
+                    .newBuilder()
+                    .setEndpoint(endpoint);
+            if (credentialsProvider != null) {
+                settings.setCredentialsProvider(credentialsProvider);
+            }
+
+            return settings.build();
+        } catch (Exception e) {
+            log.error("Error initializing Vertex AI MatchServiceSettings", e);
```

Comment:
no need to log AND rethrow an exception, please check in all places

Reply:
Done.